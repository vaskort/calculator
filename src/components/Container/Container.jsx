import React, { Component } from 'react';

import Buttons from '../Buttons/Buttons';
import Screen from '../Screen/Screen';
import calculate from '../../utilities/calculator';
import styles from './Container.module.css';

class Container extends Component {

  shouldComponentUpdate(nextProps, nextState) {
    console.log(nextState);
    return true;
  }

  constructor(props) {
    super(props);
    this.state = {
      number: [],
      queue: [],
      currentNumberOnScreen: null,
      currentValue: 0
    }
    this.handlePressNumber = this.handlePressNumber.bind(this);
    this.handleOnClear = this.handleOnClear.bind(this);
    this.handleOnPressCalculation = this.handleOnPressCalculation.bind(this);
    this.handleEquals = this.handleEquals.bind(this);
  }

  handlePressNumber(e) {
    const value = e.target.value;

    if (value === "." && this.state.number.indexOf(".") !== -1) return false;
    if (value === "0" && this.state.currentNumberOnScreen === "0") return false;

    this.state.number.push(value);
    this.setNumberOnScreen(this.state.number.join(''));
  }

  handleOnPressCalculation(e) {
    this.state.queue.push(this.state.number.join(''));
    this.state.queue.push(e.target.value);

    this.setNumberOnScreen(e.target.value);
    this.clearNumber();
  }

  handleEquals() {
    if (this.state.queue.length < 2) return;
    this.state.queue.push(this.state.number.join(''));
    const result = calculate(this.state.queue);
    this.setState({
      currentNumberOnScreen: result,
      currentValue: result,
      number: [],
      queue: [result]
    })

    this.setNumberOnScreen(result);
  }

  clearNumber() {
    this.setState({
      number: []
    })
  }

  handleOnClear() {
    this.setState({
      number: [],
      queue: [],
      currentNumberOnScreen: null,
      currentValue: null
    })
  }

  setNumberOnScreen(value) {
    this.setState({
      currentNumberOnScreen: value
    })
  }

  render() {
    const { currentNumberOnScreen } = this.state;

    return (
      <div className={styles.container}>
        <Screen value={currentNumberOnScreen} />
        <Buttons
          onPressNumber={this.handlePressNumber}
          onPressCalculation={this.handleOnPressCalculation}
          onClear={this.handleOnClear}
          onEquals={this.handleEquals}
        />
      </div>
    );
  }
}

export default Container;