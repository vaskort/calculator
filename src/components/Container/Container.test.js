import React from 'react';
import Container from './Container';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

describe('Calculator Component', () => {
  it('renders without crashing', () => {
    const wrapper = renderer.create(<Container />)

    expect(wrapper).toMatchSnapshot();
  });

  it('handlePressNumber shows the correct number on screen', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();
    const spy = jest.spyOn(wrapper.instance(), 'setNumberOnScreen');

    let e = {
      target: {
        value: '1'
      }
    }

    instance.handlePressNumber(e);
    expect(spy).toHaveBeenCalledWith(e.target.value);
  });

  it('handlePressNumber number property in the state correctly', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();

    expect.assertions(2);

    const e = {
      target: {
        value: '1'
      }
    }

    instance.handlePressNumber(e);
    expect(instance.state.number).toEqual(['1']);
    
    instance.handlePressNumber(e);
    expect(instance.state.number).toEqual(['1', '1']);
  });

  it('handleOnPressCalculation changes the state correctly', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();

    const e = {
      target: {
        value: '+'
      }
    }

    instance.setState({
      number: ['1']
    });

    instance.handleOnPressCalculation(e);

    expect(instance.state.queue).toEqual(['1', '+']);
  });

  it('handleOnPressCalculation calls setNumberOnScreen and clearNumber', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();
    const spy = jest.spyOn(wrapper.instance(), 'setNumberOnScreen');
    const spy2 = jest.spyOn(wrapper.instance(), 'clearNumber');

    expect.assertions(2);

    const e = {
      target: {
        value: '+'
      }
    }

    instance.setState({
      number: ['1']
    });

    instance.handleOnPressCalculation(e);

    expect(spy).toHaveBeenCalledWith(e.target.value);
    expect(spy2).toHaveBeenCalled();
  });

  it('handleEquals sets the properties on state correctly', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();

    expect.assertions(4);

    instance.setState({
      number: ['1'],
      queue: ['2', '+']
    });

    instance.handleEquals();

    expect(instance.state.queue).toEqual([3]);
    expect(instance.state.number).toEqual([]);
    expect(instance.state.currentNumberOnScreen).toEqual(3);
    expect(instance.state.currentValue).toEqual(3);
  });

  it('handleEquals calls setNumberOnScreen with the correct value', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();
    const spy = jest.spyOn(wrapper.instance(), 'setNumberOnScreen');

    instance.setState({
      number: ['1'],
      queue: ['2', '+']
    });

    instance.handleEquals();

    expect(spy).toHaveBeenCalledWith(3);
  });

  it('clearNumber clears the number property on state correctly', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();

    instance.setState({
      number: ['1']
    });

    instance.clearNumber();

    expect(instance.state.number).toEqual([]);
  });

  it('handleOnClear clear the state correclty', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();

    instance.setState({
      number: ['1'],
      queue: ['1'],
      currentNumberOnScreen: 1234,
      currentValue: 1234
    });

    instance.handleOnClear();

    expect(instance.state).toMatchObject({
      number: [],
      queue: [],
      currentNumberOnScreen: null,
      currentValue: null
    });
  });

  it('does not call set State to update the properties if the = is pressed with one number', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();
    const spy = jest.spyOn(wrapper.instance(), 'setState');

    instance.handleEquals();

    expect(spy).not.toHaveBeenCalled();
  });

  it('makes sure you cannot choose more than 1 zero', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();
    const spy = jest.spyOn(wrapper.instance(), 'setNumberOnScreen');
    const e = {
      target: {
        value: '0'
      }
    }

    instance.setState({
      number: ['0'],
      currentNumberOnScreen: '0',
      queue: []
    });

    instance.handlePressNumber(e);

    expect(spy).not.toHaveBeenCalledWith();
  });

  it('does not allow you to add another . if you already have one', () => {
    const wrapper = shallow(<Container />)
    const instance = wrapper.instance();

    const e = {
      target: {
        value: '.'
      }
    }

    instance.setState({
      number: ['0', '.', '0', '1'],
      queue: []
    });

    instance.handlePressNumber(e);

    expect(instance.state.number).toEqual(['0', '.', '0', '1']);
  });
});