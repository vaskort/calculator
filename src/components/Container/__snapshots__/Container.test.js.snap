// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Component renders without crashing 1`] = `
<div
  className="container"
>
  <div
    className="screen"
  />
  <div
    className="buttonsContainer"
  >
    <button
      className="button"
      onClick={[Function]}
      value="+"
    >
      +
    </button>
    <button
      className="button"
      onClick={[Function]}
      value="-"
    >
      -
    </button>
    <button
      className="button"
      onClick={[Function]}
      value="/"
    >
      /
    </button>
    <button
      className="button"
      onClick={[Function]}
      value="*"
    >
      *
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={1}
    >
      1
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={2}
    >
      2
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={3}
    >
      3
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={4}
    >
      4
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={5}
    >
      5
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={6}
    >
      6
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={7}
    >
      7
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={8}
    >
      8
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={9}
    >
      9
    </button>
    <button
      className="button"
      onClick={[Function]}
      value={0}
    >
      0
    </button>
    <button
      className="button"
      onClick={[Function]}
      value="."
    >
      .
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      Clear
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      =
    </button>
  </div>
</div>
`;
